Fix the issue where HTML elements cannot be dropped from the text selector returned by Selector.jmespath()

dream2333 opened the pull request:
When using the .xpath method to create nodes from a text type selector, it appears that these nodes are actually copies generated from the text, rather than being generated based on the original root node. As a result, when executing the .drop method, it doesn't affect the content of the original HTML tree. This issue is mostly observed when using jmespath and xpath in combination.

```
body_selector = response.jmespath("news.body")
styles = body_selector.xpath("//style")
styles.drop()
# always contains the content of the style tags
content = body_selector.xpath("string(.)").get()
```

This pull request fixes an issue where HTML elements were not being dropped correctly from a text selector. The ·_text_lazy_html_root· attribute has been added to store a temporary root node, which prevents the creation of a new root HTMLElement copy each time a text selector is used

Fixes https://github.com/scrapy/parsel/issues/297, resolves #299

dream2333 commented:
#297

Gallaecio commented:
This feels like a workaround, I wonder if there is not some root issue that needs to be addressed here. Maybe `type="text"` should be removed [here](https://github.com/scrapy/parsel/blob/740734252efd246679e5c6ee4028a47d5851712d/parsel/selector.py#L568), although it would not surprise me if that broke something else.

Could you add a test for the issue, so it is easier to experiment with alternative solutions?

dream2333 commented:
> 这感觉像是一种解决方法，我想知道这里是否没有一些需要解决的根本问题。也许应该[在这里](https://github.com/scrapy/parsel/blob/740734252efd246679e5c6ee4028a47d5851712d/parsel/selector.py#L568)删除，尽管如果这破坏了其他东西，我不会感到惊讶。`type="text"`
> 
> 您能否为该问题添加一个测试，以便更轻松地尝试替代解决方案？

Sure, I'll add a test for this issue after work. My main concern is that direct modifications to the selector might break the forward compatibility of the entire library. In the current version, the implementation of converting text to HtmlElement is not very elegant either. It doesn't cache the HtmlElement generated from text, nor does it associate the HtmlElement with the selector itself. This leads to the creation of a new copy every time a query is made on a text node. Any modifications made to this copy will not affect the original Selector at all.

dream2333 commented:
> This feels like a workaround, I wonder if there is not some root issue that needs to be addressed here. Maybe `type="text"` should be removed [here](https://github.com/scrapy/parsel/blob/740734252efd246679e5c6ee4028a47d5851712d/parsel/selector.py#L568), although it would not surprise me if that broke something else.
> 
> Could you add a test for the issue, so it is easier to experiment with alternative solutions?



> This feels like a workaround, I wonder if there is not some root issue that needs to be addressed here. Maybe `type="text"` should be removed [here](https://github.com/scrapy/parsel/blob/740734252efd246679e5c6ee4028a47d5851712d/parsel/selector.py#L568), although it would not surprise me if that broke something else.
> 
> Could you add a test for the issue, so it is easier to experiment with alternative solutions?

Sure, I'll add a test for this issue after work. My main concern is that direct modifications to the selector might break the forward compatibility of the entire library. In the current version, the implementation of converting text to HtmlElement is not very elegant either. It doesn't cache the HtmlElement generated from text, nor does it associate the HtmlElement with the selector itself. This leads to the creation of a new copy every time a query is made on a text node. Any modifications made to this copy will not affect the original Selector at all.

dream2333 committed:
Fix drop html element from a text type Selector

dream2333 committed:
Add testcases for drop html node

dream2333 committed:
Add testcases for drop html node

dream2333 committed:
Add type hint

dream2333 committed:
Fix drop html element from a text type Selector

dream2333 committed:
Merge remote-tracking branch 'origin/master'

# Conflicts:
#	parsel/selector.py
#	tests/test_selector.py

dream2333 commented:
@Gallaecio 
Updated pull request.
Added two test cases to verify the effect of dropping nodes from a Selector of type 'text'.

I tried to remove `type="text"`, but it broke jmespath. Currently, I think adding a cache for `root`'s HTML is the least destructive solution and it solves the problem of recreating etree._Element every time an xpath query is executed on a selector of type 'text'.

Gallaecio commented on parsel/selector.py:
```diff
@@ -722,25 +729,30 @@ def get(self) -> Any:
-        if self.type in ("text", "json"):
+        if self.type in ("json", "text"):
+            if self.type == "text" and self._text_lazy_html_root is not None:
+                return etree.tostring(
+                    self._text_lazy_html_root, encoding="unicode", with_tail=False
+                )
```
A problem with this approach is that we are assuming HTML, when it could be XML.

Gallaecio commented:
So, I have created #299 as an alternative, but I have no strong opinion on which way to go. To be honest, I think both could work, i.e. this approach makes things work by default for HTML, while #299 would provide a way to make things work with XML.

@kmike @wRAR Any thoughts?

AdrianAtZyte committed:
Merge remote-tracking branch 'origin/master' into dream2333

AdrianAtZyte committed:
Improve test coverage

dream2333 commented:
Is this still an active issue? I noticed this PR was reopened. I've been a heavy user of parsel for the last two years and really appreciate the tool. If this issue is still open, I'd be glad to take a look and submit a new fix for it.
@AdrianAtZyte @Gallaecio

AdrianAtZyte commented:
This PR is already *your* fix (thanks!), I have just updated it to address conflicts and improve test coverage, so that it is more likely to be merged soonish.
